python/ray/data/_internal/stats.py

```diff
@@ -258,15 +250,80 @@ def summary_string(
             else:
                 for i, metadata in stats_map.items():
                     self.stages["read"][i] = metadata[0]
+
+        stages_stats = []
+        is_substage = len(self.stages) > 1
+        for stage_name, metadata in self.stages.items():
+            stages_stats.append(
+                StageStatsSummary.from_block_metadata(
+                    metadata,
+                    self.time_total_s,
+                    stage_name,
+                    is_substage=is_substage,
+                )
+            )
+
+        iter_stats = IterStatsSummary(
+            self.iter_wait_s,
+            self.iter_get_s,
+            self.iter_next_batch_s,
+            self.iter_format_batch_s,
+            self.iter_user_s,
+            self.iter_total_s,
+        )
+        stats_summary_parents = []
+        if self.parents is not None:
+            stats_summary_parents = [p.to_summary() for p in self.parents]
+        return DatasetStatsSummary(
+            stages_stats,
+            iter_stats,
+            stats_summary_parents,
+            self.number,
+            self.dataset_uuid,
+            self.time_total_s,
+            self.base_name,
+            self.extra_metrics,
+        )
+
+
+@DeveloperAPI
+@dataclass
+class DatasetStatsSummary:
+    stages_stats: List["StageStatsSummary"]
+    iter_stats: "IterStatsSummary"
+    parents: List["DatasetStatsSummary"]
+    number: int
+    dataset_uuid: str
+    time_total_s: float
+    base_name: str
+    extra_metrics: Dict
```

Comment:
Nit: We at least know that the keys will be strings.
```suggestion
    extra_metrics: Dict[str, Any]
```